Show dragged tab images at full HiDPI resolution

eirikbakke opened the pull request:
When dragging window system tabs, show the dragged tab image at full HiDPI resolution.

Before:
<img width="682" alt="Screenshot 2025-03-11 at 12 24 19 AM" src="https://github.com/user-attachments/assets/41b61e14-bfa7-40f0-9a19-ff32fcca0403" />

After:
<img width="682" alt="Screenshot 2025-03-11 at 12 22 58 AM" src="https://github.com/user-attachments/assets/fcdff150-0a66-421e-b13b-fb2bd72e0281" />

Tested on MacOS with a Retina screen (2x scaling) and Windows 11 with 150% HiDPI scaling.

mbien commented on platform/core.windows/src/org/netbeans/core/windows/view/dnd/DragWindow.java:
```diff
@@ -44,6 +46,9 @@
+    JWindow. This ensures full-resolution painting on HiDPI and Retina screens. */
+    private static final int DPI_SCALE = 2;
 
     private boolean useFadeEffects = !Boolean.getBoolean( "winsys.dnd.nofadeeffects" );
```
the transparent drag window felt always a little slow to me. I think part of it is the fairly long fade-in effect (2s?) which makes it look less snappy as it could be. (easier to see with dark theme).

Wondering if we should turn it off by default or reduce the fade-in time?

(The second problem turned out to be my system settings which also have a small effect for opening/closing windows which clashed with this usecase)

eirikbakke replied:
Hmm, I'm not seeing much of what the fade effect is doing, on either Windows or MacOS, despite it being confirmed to be on. In general I tend to prefer to remove all animations that don't have a very specific purpose. The drag image itself is enough to show what is going on. Should I just remove the animation?

mbien replied:
maybe it was added to workaround flickering? It seems to flash white for a split second before it paints the thumbnail. Easy to spot when a dark theme is enabled and animation is off.

mbien replied:
flickering is fixed by setting the background color of the window. Tested with `setBackground(Color.RED);` which caused it to flicker red.

`JWindow` might not have look and feel color assigned?

eirikbakke replied:
Ah, now I see what the fade is doing, and I see the flickering you mention too on dark mode. But it's easy to remedy... it's just a matter of setting the window background color appropriately for the light vs. dark LAF.

I will try to make a version that rips out the complicated animation code while avoiding the flickering.

mbien replied:
while reading through the code i noticed that window transparency is implemented via JNA (DragAndDropFeedbackVisualizer). This could probably use Java 7 API but this is a bit out of scope for this PR i think.

eirikbakke replied:
I added two more proposed commits, (1) to remove the animation, and (2) to remove the flicker and one more bug I discovered.

We could try to remove the JNA stuff, but that requires testing on every platform, so I didn't do it currently.

mbien commented on platform/core.windows/src/org/netbeans/core/windows/view/dnd/DragWindow.java:
```diff
@@ -121,25 +130,40 @@ private BufferedImage createImageBuffer( BufferedImage src ) {
         return res;
     }
 
+    private static void drawImageScaled(Graphics2D g2d, Image image, int x, int y) {
+      AffineTransform oldTransform = g2d.getTransform();
```
2 space indentation here

eirikbakke replied:
Will fix!

mbien commented on platform/core.windows/src/org/netbeans/core/windows/view/dnd/DragWindow.java:
```diff
@@ -95,15 +102,17 @@ private BufferedImage createContentImage( Component c, Dimension contentSize ) {
         GraphicsConfiguration config = GraphicsEnvironment.getLocalGraphicsEnvironment()
                     .getDefaultScreenDevice().getDefaultConfiguration();
 
-        BufferedImage res = config.createCompatibleImage(contentSize.width, contentSize.height);
+        BufferedImage res = config.createCompatibleImage(
+                contentSize.width * DPI_SCALE, contentSize.height * DPI_SCALE);
         Graphics2D g = res.createGraphics();
+        g.scale(DPI_SCALE, DPI_SCALE);
```
this would be quite a large image on 4k screens. You sure its worth it to double the size to have the scaled down version look better?

eirikbakke replied:
In the default configuration, it's a small thumbnail rectangle as seen in the screenshot in the PR, not a full-size copy of the component.

But looking at DragAndDropFeedbackVisualizer, which instantiates the DragWindow, there is a WinSysPrefs.DND_SMALLWINDOWS option that can be turned off to get the full-size window. If a platform app happens to be using that setting, it's true that the buffer could be several megabytes.

Should I make some efforts to skip the scaling in the WinSysPrefs.DND_SMALLWINDOWS=false case?

mbien replied:
oh nevermind. I assumed `contentSize` is the size of the component. Its only 250x250 on 1080p so this is no problem to double it.

mbien reviewed (approved):
looks good (tested on linux/cinnamon). Nice cleanup.

I suppose lower powered systems like raspi could turn that feature off in the settings if the high quality rendering hints/buffer sizes cause problems. I vaguely remember that it had some stability issues a while ago, that is probably why I had it turned off in my workstation.

Replacing the JNA code with Java 7 window transparency API is something we could try in another cleanup at some point too.

eirikbakke commented:
Thanks for reviewing! Planning to merge without squashing since it's useful to have the "remove animation-related code" part in a separate commit.

> Replacing the JNA code with Java 7 window transparency API is something we could try in another cleanup at some point too.

I can have a look at it...

eirikbakke commented:
@mbien I ended up adding another commit that removes the use of JNA, since it was easy to do. Tested on MacOS and Windows, both with the "Transparent drag window image" option (in Options->Appearance) both enabled and disabled.

If you have a chance to test the latest version on Linux, that would cover all the OSes... thanks for your help!

mbien commented:
it doesn't work very well for me unfortunately (note the tab quality and the fact that the background rectangle isn't transparent)
![drag1](https://github.com/user-attachments/assets/bf79c8c8-4269-4a39-bd5c-bbe8a1de6a09)

Lets not let the JNA cleanup distract from the original goal, we can add a cleanup at a later point and also deprecate `NativeWindowSystem` in one go (check other usages etc).

eirikbakke commented:
@mbien Good catch; thanks for testing that! I have dropped the JNA removal commit then, restoring to the previous version you tried.

I added one small change in a separate commit, which is a rectangle fill operation that turns out to be needed on Windows in non-transparent mode. I've tested it again on Windows and MacOS, in light and dark mode, and with and without transparency enabled. I'll squash this last commit into the previous before merging.

mbien commented:
@eirikbakke please pause merges for now, see dev list.

eirikbakke commented:
@mbien Yes, saw that, will step away from the keyboard :-) Thank you for looking into it!
